### DISCUSSION TITLE
Training on Crawling @ Home

### DISCUSSION THREAD

**Discussion by afiaka87** [lucidrains/DALLE-pytorch#365] [Show and tell]
I've been running some experiments with the Crawling@Home dataset (C@H? CAH?) on DALLE. There are also some plans by others in the Discord to do so. I'll be posting some preliminary results over the next few days. 

1. [Deliverables/Code Snippets](https://github.com/lucidrains/DALLE-pytorch/discussions/365#discussioncomment-1363939)
2. [Filtering Experiment (videogame captions)](https://github.com/lucidrains/DALLE-pytorch/discussions/365#discussioncomment-1364101)
3. [512 Dim DALLE](https://github.com/lucidrains/DALLE-pytorch/discussions/365#discussioncomment-1364277)

**Comment by afiaka87** [comment]
### Deliverables/Code Snippets

I felt this code was too specific to include upstream. At any rate; if you would like to train a `DALLE-pytorch` on C@H; you can do some fun live filtering without a significant performance penalty because DALLE is (usually) the bottleneck.

```python
```python
    ENABLE_METADATA = True
    FILTER_WORDS = ["image", "photo"]
    def filter_dataset(item):
        if mycap not in item:
            return False
        current_cap = item[mycap].decode('utf-8').lower()

        # only keep captions that are long enough
        if len(item[mycap]) <= 64: return False # set minimum caption length

        # only captions with the following words are kept
        has_word = False
        for word in FILTER_WORDS: # list of words/phrases to filter out
            if word in current_cap:
                has_word = True
                break
        if not has_word: return False

        if ENABLE_METADATA: # metadata is only on C@H or img2dataset datasets
            current_metadata = json.loads(item['json'].decode('utf-8')) # json metadata

            width = int(current_metadata['original_width'])
            height = int(current_metadata['original_height'])
            similarity = float(current_metadata['similarity'])


            # only square aspect ratio
            if width != height:
                return False

            if width < IMAGE_SIZE: return False # only images larger than IMAGE_SIZE

            if similarity < 0.4: return False # only images with similarity >= 0.4

        if myimg not in item:
            return False
        
        return True
    
    w_dataset = wds.WebDataset(DATASET, handler=wds.warn_and_continue)
    filtered_dataset = w_dataset.select(filter_dataset)
    ds = filtered_dataset.map_dic
```

**Comment by afiaka87** [comment]
### Filtering Experiment (videogame captions)

Using caption-filtering; you can start training on subsets containing words right away.

For instance, filtering on various videogame related keywords finds a good deal of videogame box art. 

(cherrypicked) You can see the DALLE begins to learn the difference between xbox/ps4:
![media_images_image_11100_c248d89aacf25f844386](https://user-images.githubusercontent.com/3994972/134237921-e671d041-b3d8-4587-8753-800cdebbf39a.png)
![media_images_image_9200_96a315effba4810b02c4](https://user-images.githubusercontent.com/3994972/134237924-90c981da-d92c-4db7-9d00-f630101175af.png)

![media_images_image_16100_34f319098a4c4b1ccf4c](https://user-images.githubusercontent.com/3994972/134237931-c37bafe6-f78d-4448-8200-27c512257088.png)

A link to the full run is here:
https://wandb.ai/dalle-pytorch-replicate/CAH_VG/reports/DALLE-pytorch-on-C-H-videogame-captions---VmlldzoxMDQzNDg3?accessToken=omqfipkjpdjjif1g6kgll736hy208vjdo91erff8mtim765ldkv661r29l2rl879

**Comment by afiaka87** [comment]
### C@H for 900K Samples on 512 Dim, 128 TSL, 16 Depth, 16 Heads, Rotary, Shift

https://wandb.ai/dalle-pytorch-replicate/CAH_dalle_imagenet

**Reply by afiaka87** [reply]
These were trained with no filtering. This means that aspect-ratio can be non-square (using white for the fillcolor).

**Reply by afiaka87** [reply]
An interesting experiment to try would be to include the aspect ratio in the caption (it's available in the metadata).

**Reply by afiaka87** [reply]
My attempts to improve the loss by decreasing the learning rate were perhaps misguided - I don't think the loss had plateaued enough to need that.
